### ISSUE TITLE
Allow timers to last longer than 7 days

### ISSUE THREAD

**Issue by cgillum** [Azure/azure-functions-durable-extension#14]
## Issue
The Azure Storage provider uses invisible queue messages to implement durable timers. Queue messages have a maximum lifetime of 7 days, so durable timer messages cannot live longer than this.

## Proposal
Allow durable timers to have an arbitrary lifetime, which could go as high as weeks or months.

## Motivation
Real world scenarios often involve extended timeouts. For example, a late fee might be applied to a bill if it is not paid 30 days after it is due. A durable timer is a natural way to track this kind of expiration, and a 7 day limitation reduces the usefulness of the feature.

## Design Notes
The fix would be specific to the Azure Storage provider extension in the Durable Task Framework (the Service Bus provider has no such limits). Azure Storage cannot support a message living longer than 7 days in the queue. To work around this, if an orchestrator creates a timer for 30 days, internally a message should be tagged with metadata indicating the intended receive time (e.g. 30 days from now) but it should be enqueued for only **72 hours** (using a number small than 7 days is preferred to avoid potential data loss if the message cannot be immediately dequeued for some reason).

Once a message is received, the provider checks to see if a `ScheduledReceiveTimeUtc` metadata property exists and if it does, compares that time to the current time. If the `ScheduledReceiveTimeUtc` is within 5 minutes from the current time, the message can be processed. Otherwise, a new copy of the message should be enqueued for as many as another 72 hours (or less if the scheduled receive time is less than 72 hours away).

## Drawbacks
* If an orchestrator function is waiting on a timer for extended periods of time, then it will be periodically woken up (consuming CPU and memory) without having any real work to do.
* A single timer event can result in multiple messages being enqueued and received in Azure Storage, which means the message ID cannot be used for end-to-end correlation. The extension already defines its own correlation ID for messages and end-to-end tracing relies on this, so it's not a significant problem, but it is something to be aware of in case someone attempts to change this correlation behavior or try to do correlation purely using Storage logs.

**Comment by SimonLuckenuik** [comment]
@cgillum is there any target date for this issue? Is there a durable task framework related issue I can monitor? I guess this is low priority considering that I can reschedule the message myself.

**Comment by cgillum** [comment]
No specific date - we just need to get around to doing it. We'll definitely include it for GA, which is targeting Spring 2018, but ideally this fix will come much sooner than that.

Right now I'm keeping the durable task framework bugs logged here rather than in the DTFx repo since I'm the only consumer of the Azure Storage provider at this point.

**Comment by SimonLuckenuik** [comment]
In the meantime, I am simply creating timer with a max value below the limitation and rescheduling until the target time is reached.

**Comment by vhe1** [comment]
Why not just use ServiceBus instead of storage queues?

**Comment by cgillum** [comment]
We considered Service Bus at the beginning of this project since it's natively supported by the Durable Task Framework. The main problem was that the required features require a standard subscription ($10 USD/month). Azure Storage, on the other hand, is purely consumption-based billing, and is already part of every function app. We are considering allowing different storage providers in the future, but for now we're sticking with Azure Storage, in spite of its limitations.

**Comment by gled4er** [comment]
Hello @cgillum ,

Do you think the information [here ](https://stackoverflow.com/questions/48344332/azure-queue-max-ttl) can help us  with this feature? 

In the new version of Azure Storage .NET SDK we are able to set arbiratry expiration period as shown below:

```csharp
       //WindowsAzure.Storage 8.7
        queue.AddMessage(message); // TTL is 7 days
        queue.AddMessage(message, TimeSpan.MaxValue); // Throws The argument 'timeToLive' is larger than maximum of '7.00:00:00'

        //WindowsAzure.Storage 9.0
        queue.AddMessage(message); // TTL is 7 days
        queue.AddMessage(message, TimeSpan.MaxValue); //Expiration Time is now 31.dec 9999
```

Thank you!

**Comment by SimonLuckenuik** [comment]
@gled4er according to the [release notes](https://github.com/Azure/azure-storage-net/blob/master/changelog.txt), they only support infinite lifetime, which is probably the only reason why they changed the API (Queues: Added support for infinite TTL on queue messages.).

@cgillum I see that has been removed from GA, any chance this will be added soon? Even if my workaround works, it's a pain to do, and not obvious for the developers that are getting started with Durable Functions.

**Comment by gled4er** [comment]
Hello @SimonLuckenuik,

Thank you for the comment.

I haven't tried the new TTL options yet but based on this  [documentation page](https://docs.microsoft.com/en-us/rest/api/storageservices/put-message) I think you can set any int value now. In my opinion infinite lifetime is only one possible option, not the only one allowed.

Thank you!

**Comment by cgillum** [comment]
@SimonLuckenuik We had to punt this out of GA for schedule/resourcing reasons, unfortunately. We're still motivated to get this done soon, though - lots of people asking for it.

I'll take a look at the documented storage updates. Arbitrary lifetime seems like a pretty big deal! The only problem is that Azure Functions is fixed at a particular Storage SDK version, so I'm not sure if we'll be able to take advantage of it (yet)...

**Comment by gled4er** [comment]
Hello @cgillum,

Did you have time to check if we will be able to use the latest Azure Storage Nuget package or we will need to come up with custom logic to support this feature?

Thank you!

**Comment by cgillum** [comment]
Yes, I confirmed that Azure Functions v2 will be upgraded to a much more recent version of the Azure Storage SDK before it goes GA, so we will be able to take advantage of this feature once that happens.

**Comment by gled4er** [comment]
Great!

**Comment by markheath** [comment]
Here's a quick hack to create an extension method that could be used to extend the capacity of the timer to an arbitrarily long time. Basically we create a Timer for the maximum period (72 hours as suggested above) and when we wake up, check if we need to keep waiting for longer by starting another CreateTimer:

        public static Task CreateTimerExtended(this DurableOrchestrationContext ctx, CancellationToken token, TimeSpan timeout)
        {
            var completeTimeout = ctx.CurrentUtcDateTime + timeout;
            var tcs = new TaskCompletionSource<int>();
            void WaitUntil()
            {
                var maxTimerDuration = TimeSpan.TimeSpan.FromHours(72);
                var timeoutAt = ctx.CurrentUtcDateTime + (timeout > maxTimerDuration ? maxTimerDuration : timeout);

                var timeoutTask = ctx.CreateTimer(timeoutAt, token);
                timeoutTask.ContinueWith(t =>
                {
                    if (t.IsCanceled)
                        tcs.TrySetCanceled();
                    else if (t.IsFaulted)
                        tcs.TrySetException(t.Exception);
                    else if (ctx.CurrentUtcDateTime > completeTimeout)
                        tcs.TrySetResult(0);
                    else
                        WaitUntil();
                }, TaskContinuationOptions.ExecuteSynchronously);
            }
            WaitUntil();
            
            return tcs.Task;
        }

I've done some very simple tests (setting a `maximumTimerDuration` of 1 minute is a nice easy way to test this) and it seems to work OK. I haven't tested cancellation support yet.

The downside of this approach is that the orchestrator keeps waking up, so for a very long wait (e.g. > 1 month), you'd start to build up a substantial event sourcing history.

One possible way to work around that is to create this extended timer as a sub-orchestration. That way, its just the sub-orchestration that has a growing event source history that doesn't pollute the history of the main orchestration. Here's how a sub-orchestration could implement a long wait

        [FunctionName("LongWait")]
        public static async Task LongWait(
            [OrchestrationTrigger] DurableOrchestrationContext ctx,
            TraceWriter log)
        {
            var timeout = ctx.GetInput<TimeSpan>();
            var completeTimeout = ctx.CurrentUtcDateTime + timeout;
            var maxTimerDuration = TimeSpan.FromHours(72);

            using (var cts = new CancellationTokenSource())
            {
                while (ctx.CurrentUtcDateTime < completeTimeout)
                {
                    var timeoutAt = ctx.CurrentUtcDateTime + (timeout > maxTimerDuration ? maxTimerDuration : timeout);
                    await ctx.CreateTimer(timeoutAt, cts.Token);
                }
            }
        }

You might even be able to use `ContinueAsNew` in this sub-orchestrator to eliminate the need for a growing event sourcing history altogether.

The downside of the sub-orchestrator approach is that there doesn't seem to be a way to pass a cancellation token to a sub-orchestration. But if there is a way to do that, I wonder if this is a quicker way to get this feature without needing to put metadata on the queue messages.

**Comment by Expecho** [comment]
Any estimation on when this issue will be resolved?

**Comment by RockyBalboa2018** [comment]
Any update on this issue ?

**Comment by cgillum** [comment]
We're tracking this for the final release of v2.0, which we hope to have released later this year. Until then, I suggest looking into the workarounds mentioned by Mark above.

**Comment by underscoreHao** [comment]
Any update on this issue?

**Comment by cgillum** [comment]
Unfortunately we had to punt this out past the initial v2.0 release due to time constraints. We're still looking into getting this done soon, however, since it's a common ask.

**Comment by ConnorMcMahon** [comment]
I was hoping to address this with [this PR](https://github.com/Azure/azure-functions-durable-extension/pull/1100), but unfortunately the improvements from the Azure Storage SDK in 8.x vs 9.x only improved TTL, not visibility delays. 

I think the best approach for this work item is for us to internally handle timers in a way similar to @markheath's workaround. This unfortunately will no longer make it into 2.1 because of that.

**Comment by jonasgranlund2** [comment]
Hi Chris and Connor, I would like to create long running Scheduled Durable Entity events with the ctx.SignalEntity method that take the dateTime overload for scheduling.  Since it is the same 7 days max-time on those I was wondering if you have any recomended pattern? It would be great to create something generic but since the SignalEntity is fire and forget I dont succeed to come up with a good solution. I have tried Mark Hats solution with Extension method and it works gret, but in this scenario I would like it to be more light weight than to create a new orchstration instance per incoming event.
